### PR TITLE
Fix `instancecheck`, `subclasscheck` of `UnionType`

### PR DESCRIPTION
**Pull request by TwoPair**
Resolve `test_types::UnionTests::test_instancecheck, test_bad_instancecheck,test_bad_subclasscheck`

### CODE REVIEW

**Comment by youknowone** [comment]
@Snowapril could you review this PR?

File: vm/src/builtins/union.rs [hunk 1]
```diff
@@ -90,15 +90,13 @@ impl PyUnion {
...
+        obj.is_instance(zelf.args().as_object(), vm)
     }
 
     #[pymethod(magic)]
-    fn subclasscheck(_zelf: PyRef<Self>, _obj: PyObjectRef, vm: &VirtualMachine) -> PyResult {
-        Err(vm
-            .new_type_error("issubclass() argument 2 cannot be a parameterized generic".to_owned()))
+    fn subclasscheck(zelf: PyRef<Self>, obj: PyObjectRef, vm: &VirtualMachine) -> PyResult<bool> {
+        obj.is_subclass(zelf.args().as_object(), vm)
     }
```
Comment (Snowapril) [inline vm/src/builtins/union.rs:100]: We need to check whether `GenericAlias` object exist in args and then return `TypeError` with message "isinstance() argument 2 cannot contain a parameterized generic"

File: vm/src/builtins/union.rs [hunk 2]
```diff
@@ -90,15 +90,13 @@ impl PyUnion {
...
+    fn instancecheck(zelf: PyRef<Self>, obj: PyObjectRef, vm: &VirtualMachine) -> PyResult<bool> {
+        obj.is_instance(zelf.args().as_object(), vm)
     }
 
     #[pymethod(magic)]
-    fn subclasscheck(_zelf: PyRef<Self>, _obj: PyObjectRef, vm: &VirtualMachine) -> PyResult {
-        Err(vm
-            .new_type_error("issubclass() argument 2 cannot be a parameterized generic".to_owned()))
+    fn subclasscheck(zelf: PyRef<Self>, obj: PyObjectRef, vm: &VirtualMachine) -> PyResult<bool> {
+        obj.is_subclass(zelf.args().as_object(), vm)
```
Comment (Snowapril) [inline vm/src/builtins/union.rs:99]: We need to check whether `GenericAlias` object exist in args and then return `TypeError` with message "isinstance() argument 2 cannot contain a parameterized generic"

```python
>>>>> isinstance(1, int | list[str])
# rustpython
True
# cpython
TypeError: isinstance() argument 2 cannot contain a parameterized generic
```

Comment (Snowapril) [reply vm/src/builtins/union.rs:99]: Hmm... As `is_instance` and `is_subclass` in the `object.rs` are already checking each argument in the tuple, so we don't have to iterate args and call `is_instance` or `is_subclass`. I also agree with your implmenetation.

But in cpython, they switched to the implementation that iterates all the args in `union` and call `is_instance` or `is_subclass` for each one. (https://github.com/python/cpython/pull/27120)

I don't know the exact reason yet, haha.. 😂

Comment (TwoPair) [reply vm/src/builtins/union.rs]: Thx for recognizing me!

**Comment by TwoPair** [comment]
I made `test_types.py` update by referring to [CPython 3.10's one](https://github.com/python/cpython/blob/0579a187df9ca5d1bd928eec710c29aa40acd826/Lib/test/test_types.py#L704-L759) for more precise checks.

**Review by Snowapril — APPROVED** [review]

**Comment by youknowone** [comment]
The changes looks great, thank you! Just one more request. Could you rebase this PR to 2 commits - one for test changes and the other one for non-test code changes?
When I'd going to squash all commits to single commit, I usually can do it my self with github UI, but not when I wants more than 2 commits.

**Review by youknowone — APPROVED** [review]
Great! Thank you!

**Pull request merged**